docs/typescript/reference/dbos-class.md

```diff
@@ -0,0 +1,957 @@
+```typescript
+DBOS.launch(httpApps?: DBOSHttpApps)
+DBOS.shutdown()
+```
+
+* `launch`: This starts the DBOS app, or registers it to be started in conjunction with a provided HTTP server.  Launching will initialize all DBOS components, run any application initializers, start workflow recovery, start event processing, and start HTTP services (if so configured).
+* `shutdown`: This stops the DBOS app.  First, handling of new events and requests is disabled, and then the workflow executor and database connections are destroyed.
+
+### Recovering Workflows
+DBOS generally recovers workflows automatically.  However, this process can be initiated externally.
+
+```typescript
+DBOS.recoverPendingWorkflows(executorIDs: string[] = ["local"]): Promise<WorkflowHandle<unknown>[]>
+DBOS.executeWorkflowById(workflowId: string, startNewWorkflow: boolean = false): Promise<WorkflowHandle<unknown>>
+```
+
+* `DBOS.recoverPendingWorkflows`: Retrieves a list of workflows to recover, and starts them, returning their handles
+* `DBOS.executeWorkflowById`: Starts executing a workflow given its `workflowId`.  If `startNewWorkflow` is `true`, the workflow is cloned and assigned a new unique workflow ID.
+
+## Extensibility
+DBOS Transact allows libraries, such as those designed to receive external events, to initiate DBOS functions.
+
+### Decorators
+Such libraries may choose to implement method decorators.  If the method decorator is registering a function with DBOS Transact, it should call `DBOS.registerAndWrapContextFreeFunction` for functions that do not accept a context as the first parameter.
+
+```typescript
+DBOS.registerAndWrapContextFreeFunction<This, Args extends unknown[], Return>(
+    target: object,
+    propertyKey: string,
+    descriptor: TypedPropertyDescriptor<(this: This, ...args: Args) => Promise<Return>>,
+)
+```
+
+`DBOS.registerAndWrapContextFreeFunction` registers the function named `propertyKey` on the `target` class, creates a wrapper function, and updates the `descriptor` to use the wrapper function.
+
+### Calling DBOS Functions From Event Dispatchers
+DBOS functions that are decorated as `DBOS.workflow`, `DBOS.transaction`, or `DBOS.step`, can be run directly from event receiver dispatch loops.
+
+For more complex circumstances, such as invoking functions with a context argument, `DBOS.executor` provides access to the [`DBOSExecutorContext` interface](../reference/contexts.md#dbosexecutorcontext).
+
+#### Setting Authenticated User And Roles
+When calling DBOS functions, it may be desired to set context variables, such as the authenticated user, and the current tracing span.  The following functions can be used for these purposes:
+
+```typescript
+// Set the authenticated user, and roles allowed by the authentication system for the scope of the callback() function
+DBOS.withAuthedContext<R>(authedUser: string, authedRoles: string[], callback: () => Promise<R>): Promise<R>
```

Comment:
I think instead of having separate contexts, can we create a single context type with the auth, trace, and caller name fields? Then, it will also allow users to attach their custom fields into the context object.

Reply:
Could possibly be added.

Reply:
Also see comment above about needing more real experience with it before the API and doc will be very good.

Reply:
So, the idea here was that these augment any existing context in certain ways, or start a new one.  We could allow use of `runWithTopContext`, this does what you want but isn't exported from `DBOS` right now.

Reply:
Not a priority, honestly let's just remove the entire "Extensibility" section and come back to it later.